Simple CRL server using native openssl CRL capabilities

jni2000 opened the pull request:
This is a Python implementation of simple CRL server (a.k.a CDP) using native openssl CRL capabilities for test cases like SCS/SDS/SSS.11, SCS/SDS/SSS.16. This server is a light-weighted but fully capable CRL server implemented using Python. The server can be deployed and run independently or can be included in the test codes run -- depends on final decision. 

This is the proposed solution for providing such CRL server capabilities to the ITS in supporting the corresponding test runs. Please review.

A reference of the openssl CRL capabilities can be found in https://jamielinux.com/docs/openssl-certificate-authority/certificate-revocation-lists.html

jni2000 commented:
Ok, will do. Thanks.

kate-harrison commented:
Some other details that came to mind:
 - We should include a script which actually starts the CRL server, not just the class.
 - Presumably there is some amount of authentication required so that the SAS UUT can be sure that this is the right CRL server (i.e. to avoid spoofing attacks). How will this be handled? We should match whatever is the standard approach used by CAs.

pmarks-net commented on src/harness/simple_crl_server.py:
```diff
@@ -0,0 +1,105 @@
+
+  def __init__(self, host_name, port, crl_url, crl_file_name):
+    """
+    Constructor for  CRL test harness Requests from SAS Under Test
```
Please search this file for "  " (double spaces) and remove the redundant ones.

jni2000 replied:
This   comment has been addressed now.

pmarks-net commented on src/harness/simple_crl_server.py:
```diff
@@ -0,0 +1,105 @@
+    """
+    Constructor for  CRL test harness Requests from SAS Under Test
+    Args:
+        "host_name": Entity hosting the  crl server
```
Use consistent formatting here (remove the quotes and excess whitespace).

For example: https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/blob/master/src/harness/sas.py#L72

jni2000 replied:
This   comment has been addressed now.

pmarks-net commented on src/harness/simple_crl_server.py:
```diff
@@ -0,0 +1,105 @@
+        "port"    : Port number at which  crl server will run
+        "crl_url" :  Name of the crl url to be validated
+        "crl_file_name" : Name of the crl file
+    Returns
```
Delete "Returns" if nothing is returned.

jni2000 replied:
Doc   string empty returns  is removed now

pmarks-net commented on src/harness/simple_crl_server.py:
```diff
@@ -0,0 +1,105 @@
+    """
+    self.server.shutdown()
+    logging.info('Stopped CRL Server at %s',
+                 self.host_name+":" + str(self.port))
```
Use consistent spaces.

jni2000 replied:
This   comment has been addressed now.

pmarks-net commented on src/harness/simple_crl_server.py:
```diff
@@ -0,0 +1,105 @@
+
+
+class CrlHttpServer(HTTPServer):
+  """ The class CrlHttpServer overrides built-in HTTPServer and takes the
```
Delete leading space.

jni2000 replied:
This   comment has been addressed now.

pmarks-net commented on src/harness/simple_crl_server.py:
```diff
@@ -0,0 +1,105 @@
+    self.log_message('Received GET Request:{}'.format(parsed_path.path))
+    if parsed_path.path == self.server.crl_url:
+      # Handles the GET Requests
+      if not os.path.exists(self.server.crl_file_name):
```
Instead of pre-testing the file, call `open()` from a `try: / except IOError:` block.  Otherwise, there's a potential race.

jni2000 replied:
Enhanced   now check the file path present if     not return 404 response

pmarks-net commented on src/harness/simple_crl_server.py:
```diff
@@ -0,0 +1,105 @@
+
+
+class CrlServerTestHarness(threading.Thread):
+  """
```
Docstrings should not begin with a newline.  See https://www.python.org/dev/peps/pep-0257/

jni2000 replied:
Changed   accordingly

jni2000 commented:
Kate and Piotr

We have addressed your general and code-specific comments and made necessary changes, please give an check and help speeding up the remaining SCS/SDS/SSS test cases already due. Thanks.

pmarks-net commented on src/harness/simple_crl_server.py:
```diff
@@ -0,0 +1,308 @@
+      return
+
+    # If the crl_file_path does not exist then respond with 404 response.
+    if not os.path.exists(self.server.crl_file_path):
```
There is still a potential filesystem race between `exists()` and `open()`.  Just call `open()` and handle the error.

See https://docs.python.org/2/glossary.html#term-eafp

jni2000 replied:
Followed   the EAFP coding style now.

pmarks-net commented on src/harness/simple_crl_server.py:
```diff
@@ -0,0 +1,308 @@
+
+  # Update CRL URL in openssl.cnf.
+  crl_url = 'URI.0 = {}'.format(DEFAULT_CRL_URL)
+  update_crl_section_command = 'sed -i -e \'s~^URI.0\\s=\\s.*$~{}\'~1 ../../cert/openssl.cnf'.\
```
Use a `r"raw string"` to avoid excess backslashes.

PEP8 says "Long lines can be broken over multiple lines by **wrapping expressions in parentheses**. These should be used **in preference to using a backslash for line continuation**."

jni2000 replied:
Changed   accordingly

pmarks-net commented on src/harness/simple_crl_server.py:
```diff
@@ -0,0 +1,308 @@
+    """This method is used to stop HTTPServer Socket"""
+
+    self.server.shutdown()
+    logging.info('Stopped CRL Server at %s', self.host_name+":" + str(self.port))
```
Use consistent spacing around `":"`

jni2000 replied:
Fixed.

pmarks-net commented on cert/openssl.cnf:
```diff
@@ -481,8 +486,9 @@ policyIdentifier = 2.16.840.1.114412.2.1
 subjectKeyIdentifier = hash
 basicConstraints = CA:FALSE
-keyUsage = critical, digitalSignature, keyEncipherment
+keyUsage = critical, digitalSignature, keyEncipherment,cRLSign
```
Use consistent spacing after `,`

jni2000 replied:
changed

pmarks-net commented on cert/openssl.cnf:
```diff
@@ -496,9 +502,10 @@ policyIdentifier = 2.16.840.1.114412.2.1
 authorityKeyIdentifier = keyid:always,issuer
 basicConstraints = CA:FALSE
-keyUsage = critical, digitalSignature, keyEncipherment
+keyUsage = critical, digitalSignature, keyEncipherment,cRLSign
```
Use consistent spacing after `,`

jni2000 replied:
changed

kate-harrison commented:
I just wanted to quickly confirm that Paul is indeed a colleague of mine.

jni2000 commented:
Sorry, Paul, I though the comments were from Piotr. Thanks and welcome aboard. Are you a colleague of Kate at Google ? :)

jni2000 commented on src/harness/certs/revoke_and_generate_crl.sh:
```diff
@@ -0,0 +1,57 @@
+elif [ "$1" == "-u"  ]; then
+    generate_crl_chain
+else
+    echo "wrong option other than (CBSD,DP,SAS,-r)"
```
fixed now

jni2000 commented on src/harness/certs/revoke_and_generate_crl.sh:
```diff
@@ -0,0 +1,57 @@
+#   Argument1: Name of the certificate to be  blacklisted
+#   Argument2: Intermediate CA to generate CRL.
+    openssl ca -revoke $1 -keyfile private/$2.key -cert $2.cert \
+    -config ../../../cert/openssl.cnf
```
Added 4   space indent for multi line code.

jni2000 commented on src/harness/certs/revoke_and_generate_crl.sh:
```diff
@@ -0,0 +1,57 @@
+#Create a CRL for root CA containing the revoked intermediate CA certificate
+echo "\n\n Generate CRL for root_ca"
+openssl ca -gencrl -keyfile private/root_ca.key -cert root_ca.cert \
+     -config ../../../cert/openssl.cnf  -crlhours 1\
```
align   with 4 space indentation now.

jni2000 commented on src/harness/certs/revoke_and_generate_crl.sh:
```diff
@@ -0,0 +1,57 @@
+}
+
+# Argument1 : Type (CBSD,DP,SAS,-u)
+if [ "$1" == "CBSD"  ]; then
```
Removed   all double space wherever applicable

jni2000 commented on src/harness/simple_crl_server.py:
```diff
@@ -0,0 +1,308 @@
+    # Setting the parameters for handler class
+    logging.info("Started CRL Server at %s", self.host_name + ":" + str(self.port))
+    self.server = CrlHttpServer(self.crl_url_path, self.crl_file_path, (self.host_name, self.port),
+                                 CrlServerHttpHandler)
```
Fixed.

jni2000 commented on src/harness/simple_crl_server.py:
```diff
@@ -0,0 +1,308 @@
+  """
+
+  def __init__(self, crl_url, crl_file_path):
+    """ Constructor for simple CRL Server.
```
Doc   string is updated correctly.

jni2000 commented:
Paul

We have addressed all your review comments and made necessary changes. Please move on and approve to unblock the related security test cases if no more concerns. Thanks.

pmarks-net commented on src/harness/certs/revoke_and_generate_crl.sh:
```diff
@@ -4,54 +4,51 @@ mkdir -p crl
-elif [ "$1" == "-u"  ]; then
-    generate_crl_chain
+#Argument1 : Type (CBSD,DP,SAS,-u)
+if [ $1 == "CBSD" ]; then
```
`"$1"` (and ideally, all shell variables) must be quoted:

    $ ./revoke_and_generate_crl.sh
    ./revoke_and_generate_crl.sh: line 43: [: ==: unary operator expected
    ./revoke_and_generate_crl.sh: line 45: [: ==: unary operator expected
    ./revoke_and_generate_crl.sh: line 47: [: ==: unary operator expected
    ./revoke_and_generate_crl.sh: line 49: [: ==: unary operator expected
    Wrong option other than (CBSD,DP,SAS,-u)

jni2000 replied:
Thanks, fixed all places.

jni2000 commented on src/harness/simple_crl_server.py:
```diff
@@ -195,21 +198,24 @@ def revokeCertificate():
   crl_url = 'URI.0 = {}'.format(DEFAULT_CRL_URL)
-  update_crl_section_command = 'sed -i -e \'s~^URI.0\\s=\\s.*$~{}\'~1 ../../cert/openssl.cnf'.\
-    format(crl_url)
+  update_crl_section_command = (r"sed -i -e 's~^URI.0\s=\s.*$~%s'~1 ../../cert/openssl.cnf"
```
thanks, fixed now.

jni2000 commented:
Paul

Thanks for the latest two review comments. We have fixed/changed as you pointed out. Please take a look.

James

pmarks-net commented on src/harness/simple_crl_server.py:
```diff
@@ -0,0 +1,320 @@
+    # with 404 'Not Found' response.
+    try:
+      with open(os.path.join(self.server.crl_file_path)) as file_handle:
+        self.wfile.write(file_handle.read())
```
When I try to fetch http://localhost:9007/ca.crl from Chrome, it fails with ERR_INVALID_HTTP_RESPONSE, because the data arrives before the headers.  I recommend this pattern:

    try:
      with open(os.path.join(self.server.crl_file_path)) as file_handle:
        file_data = file_handle.read()
    except IOError:
      ...
    else:
      ...
      self.end_headers()
      self.wfile.write(file_data)

(This also fixes the edge case where a `write()` error could trigger a 404 response.)

jni2000 replied:
Thanks   for catching this. We tested using 'wget' and it was working. We see that it   is not working with chrome. The code has been fixed as suggested.

jni2000 commented:
Paul

We made necessary changes as you suggested and fixed the chrome access problem. Please take a look and move forward to allow unblocking other SCS/SDS/SSS test cases.

Thanks.

pmarks-net commented on src/harness/certs/revoke_and_generate_crl.sh:
```diff
@@ -0,0 +1,54 @@
+function generate_crl_chain()
+{
+  #Create a CRL for root CA containing the revoked intermediate CA certificate.
+  echo "\n\n Generate CRL for root_ca"
```
These need `echo -e` for the formatting to work:

    $ echo "\ntest"
    \ntest
    $ echo -e "\ntest"
    
    test

(Also update the corresponding `\n`s in CRLServer-README.md)

jni2000 replied:
thanks, fixed all places.

pmarks-net commented on src/harness/simple_crl_server.py:
```diff
@@ -0,0 +1,331 @@
+
+  try:
+    crl_server = SimpleCrlServer(DEFAULT_CRL_URL, os.path.join(getCertsDirectoryAbsolutePath(),
+                                                             'crl/ca.crl'))
```
Indent +2 spaces, to preserve alignment.

jni2000 replied:
fixed. thanks.

jni2000 commented:
Thanks, Kate and Paul. Have a great weekend.

On Mar 23, 2018 3:29 PM, Kate <notifications@github.com> wrote:

@kate-harrison approved this pull request.

Paul does not yet have the right GitHub permissions (I'm fixing that), so I am re-approving on his behalf.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/pull/469#pullrequestreview-106622649>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AKpZrC6BM9iGG_Gm7ZztC-5zrIPXmKKBks5thU0FgaJpZM4SlZd_>.

jni2000 commented:
Thanks, Paul. We have fixed the echo issue and alignment you pointed out.
